### ISSUE TITLE
pidprofile does not create FIO and ROR during replication if empty/None on original side

### ISSUE THREAD

**Issue by cookie33** [EUDAT-B2SAFE/B2SAFE-core#81]
Testing shows:
```
$ ./testB2SafeCmd.py -test b2safe
Test b2safe
Test : basic iRODS functions. ... ok
Test : put a file in iRODS ... ok
Test : search for non existing handles ... ok
Test : create a PID using b2safe (ror=None, fio=None, fixed_content=false) ... ok
Test : create a PID using b2safe (ror=pid,  fio=None, fixed_content=false) ... ok
Test : create a PID using b2safe (ror=None, fio=pid,  fixed_content=false) ... ok
Test : create a PID using b2safe (ror=None, fio=None, fixed_content=true) ... ok
Test : create a PID using b2safe (ror=pid,  fio=pid,  fixed_content=true) ... ok
Test : create a PID using b2safe (ror=841/test,  fio=841/test,  fixed_content=false) ... ok
Test : create PIDs for a collection and files using b2safe (fixed_content=false) ... ok
Test : create PIDs for a collection and files using b2safe (fixed_content=true) ... ok
Test : create PIDs for a collection and files using b2safe using similar directory names (fixed_content=false) ... ok
Test : create PIDs for a collection and files using b2safe using similar directory names (fixed_content=true) ... ok
Test : replicate a single file locally using b2safe (not registered, non recursive) ... ok
Test : replicate a single file locally using b2safe (not registered,     recursive) ... ok
Test : replicate a single file locally using b2safe (    registered, non recursive, ror=pid, fio=pid, fixed_content=false) ... ok
Test : replicate a single file locally using b2safe (    registered, non recursive, ror=841/test, fio=None, fixed_content=false) ... FAIL
Test : replicate a single file locally using b2safe (    registered, non recursive, ror=None, fio=None, fixed_content=false) ... FAIL
Test : replicate a single file locally using b2safe (    registered,     recursive, ror=pid, fio=pid, fixed_content=false) ... ok
Test : replicate a single directory locally using b2safe (not registered,    recursive) ... ok
Test : replicate a single directory locally using b2safe (    registered,    recursive, ror=pid, fio=pid, fixed_content=false) ... ok
Test : replicate multiple directories locally using b2safe (not registered,    recursive) ... ok
Test : replicate multiple directories locally using b2safe (    registered,    recursive, ror=pid, fio=pid, fixed_content=false) ... ok

======================================================================
FAIL: Test : replicate a single file locally using b2safe (    registered, non recursive, ror=841/test, fio=None, fixed_content=false)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/robertv/git/B2SAFE-core/scripts/tests/testB2SafeCmd/irodsb2safetest.py", line 979, in test_70_b2safe_local_for_one_file_registered_non_recursive_02
    test_assert_array(self, test_values)
  File "/home/robertv/git/B2SAFE-core/scripts/tests/testB2SafeCmd/irodsb2safetest.py", line 56, in test_assert_array
    self.assertEqual(result, expected, output_string)
AssertionError: 'None' != '21.T12996/3fef9488-1529-11e7-8f8c-5254000df0ed' : EUDAT/FIO has no value, we expect 21.T12996/3fef9488-1529-11e7-8f8c-5254000df0ed

======================================================================
FAIL: Test : replicate a single file locally using b2safe (    registered, non recursive, ror=None, fio=None, fixed_content=false)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/robertv/git/B2SAFE-core/scripts/tests/testB2SafeCmd/irodsb2safetest.py", line 1053, in test_70_b2safe_local_for_one_file_registered_non_recursive_03
    test_assert_array(self, test_values)
  File "/home/robertv/git/B2SAFE-core/scripts/tests/testB2SafeCmd/irodsb2safetest.py", line 56, in test_assert_array
    self.assertEqual(result, expected, output_string)
AssertionError: 'None' != '21.T12996/46152eb8-1529-11e7-951a-5254000df0ed' : EUDAT/FIO has no value, we expect 21.T12996/46152eb8-1529-11e7-951a-5254000df0ed

----------------------------------------------------------------------
Ran 23 tests in 226.539s

FAILED (failures=2)
````
The logfiles show:
```
Mar 30 11:14:16 pid:27935 NOTICE: writeLine: inString = INFO: Starting test: test_70_b2safe_local_for_one_file_registered_non_recursive_02
...
vNOTICE: writeLine: inString = INFO: [EUDATePIDcreate] Create PID with extratype parameter: EUDAT/ROR=841/test;EUDAT/FIXED_CONTENT=False;EUDAT/PROFILE_VERSION=1;EUDAT/CHECKSUM=sha2:f4OxZX/x/FO5LcGBSKHWXfwtSx+j1ncoSt3SABJtkGk=;EUDAT/CHECKSUM_TIMESTAMP=2017-03-30T09:14:16Z
...
Mar 30 11:14:17 pid:27959 NOTICE: writeLine: inString = INFO: EUDATiFieldVALUEretrieve -> looking for EUDAT/FIO of /bob/home/robertv/test_b2safe_data1.txt
Mar 30 11:14:17 pid:27959 NOTICE: writeLine: inString = DEBUG: [EUDATSearchAndDefineField] got EUDAT/FIO=None from iCAT
...
Mar 30 11:14:17 pid:27959 NOTICE: writeLine: inString = DEBUG: [EUDATPIDRegistration] the path /bob/home/robertv/test_b2safe_data2.txt has EUDAT/PARENT=21.T12996/3fef9488-1529-11e7-8f8c-5254000df0ed, EUDAT/ROR=841/test, EUDAT/FIO=None, EUDAT/FIXED_CONTENT=False
...
Mar 30 11:14:17 pid:27959 NOTICE: writeLine: inString = INFO: [EUDATePIDcreate] Create PID with extratype parameter: EUDAT/ROR=841/test;EUDAT/PARENT=21.T12996/3fef9488-1529-11e7-8f8c-5254000df0ed;EUDAT/FIXED_CONTENT=False;EUDAT/PROFILE_VERSION=1;EUDAT/CHECKSUM=sha2:f4OxZX/x/FO5LcGBSKHWXfwtSx+j1ncoSt3SABJtkGk=;EUDAT/CHECKSUM_TIMESTAMP=2017-03-30T09:14:17Z
```
It has 2 errors:
* It creates an AVU  `EUDAT/FIO=None` on the master side. There should have been None.
* it does NOT create an AVU `EUDAT/FIO` and handle entry for the replica.

**Comment by cookie33** [comment]
This happens here:
```
EUDATPIDRegistration(*source, *destination, *notification, *registration_response) {

    logInfo("[EUDATPIDRegistration] registration of PIDs for replication from *source to *destination");

    *registration_response = "None";
    *parentPID = "None";
    *parentROR = "None";
    *childPID = "None";

    EUDATGetZoneNameFromPath(*destination, *zoneName);
    EUDATGetZoneHostFromZoneName(*zoneName, *zoneConn);
    logDebug("[EUDATPIDRegistration] remote zone name: *zoneName, connection contact: *zoneConn");

    EUDATSearchAndCreatePID(*source, *parentPID);
    if (*parentPID == "empty" || (*parentPID == "None")) {
        *registration_response = "PID of source *source is None";
        logDebug(*registration_response);
        # Update Logging (Statistic File and Failed Files)
        EUDATUpdateLogging(bool("false"),*source,*destination,*registration_response);
    }
    else {
        *parentROR = EUDATSearchAndDefineField(*source, *parentPID, "EUDAT/ROR");
        *fio = EUDATSearchAndDefineField(*source, *parentPID, "EUDAT/FIO");
        *fixed = EUDATSearchAndDefineField(*source, *parentPID, "EUDAT/FIXED_CONTENT");
        logDebug("[EUDATPIDRegistration] the path *destination has EUDAT/PARENT=*parentPID, "
                 ++ "EUDAT/ROR=*parentROR, EUDAT/FIO=*fio, EUDAT/FIXED_CONTENT=*fixed");
```
The function:
```
  *fio = EUDATSearchAndDefineField(*source, *parentPID, "EUDAT/FIO");
```
gives back `None`.

That is OK. But than it should have taken the PID value and used that for `EUDAT/FIO`.

**Comment by cookie33** [comment]
2 fixes in code done and tested:
```
$ ./testB2SafeCmd.py -test b2safe
Test b2safe
Test : basic iRODS functions. ... ok
Test : put a file in iRODS ... ok
Test : search for non existing handles ... ok
Test : create a PID using b2safe (ror=None, fio=None, fixed_content=false) ... ok
Test : create a PID using b2safe (ror=pid,  fio=None, fixed_content=false) ... ok
Test : create a PID using b2safe (ror=None, fio=pid,  fixed_content=false) ... ok
Test : create a PID using b2safe (ror=None, fio=None, fixed_content=true) ... ok
Test : create a PID using b2safe (ror=pid,  fio=pid,  fixed_content=true) ... ok
Test : create a PID using b2safe (ror=841/test,  fio=841/test,  fixed_content=false) ... ok
Test : create PIDs for a collection and files using b2safe (fixed_content=false) ... ok
Test : create PIDs for a collection and files using b2safe (fixed_content=true) ... ok
Test : create PIDs for a collection and files using b2safe using similar directory names (fixed_content=false) ... ok
Test : create PIDs for a collection and files using b2safe using similar directory names (fixed_content=true) ... ok
Test : replicate a single file locally using b2safe (not registered, non recursive) ... ok
Test : replicate a single file locally using b2safe (not registered,     recursive) ... ok
Test : replicate a single file locally using b2safe (    registered, non recursive, ror=pid, fio=pid, fixed_content=false) ... ok
Test : replicate a single file locally using b2safe (    registered, non recursive, ror=841/test, fio=None, fixed_content=false) ... ok
Test : replicate a single file locally using b2safe (    registered, non recursive, ror=None, fio=None, fixed_content=false) ... ok
Test : replicate a single file locally using b2safe (    registered,     recursive, ror=pid, fio=pid, fixed_content=false) ... ok
Test : replicate a single directory locally using b2safe (not registered,    recursive) ... ok
Test : replicate a single directory locally using b2safe (    registered,    recursive, ror=pid, fio=pid, fixed_content=false) ... ok
Test : replicate multiple directories locally using b2safe (not registered,    recursive) ... ok
Test : replicate multiple directories locally using b2safe (    registered,    recursive, ror=pid, fio=pid, fixed_content=false) ... ok
Test : replicate multiple directories locally using b2safe (    registered,    recursive, ror=841/test, fio=None, fixed_content=false) ... ok
Test : replicate multiple directories locally using b2safe (    registered,    recursive, ror=None, fio=None, fixed_content=false) ... ok

----------------------------------------------------------------------
Ran 25 tests in 314.091s

OK
```

**Comment by cookie33** [comment]
see pull request #78 for a fix..

**Comment by ccacciari** [comment]
addressed in b2safe version 4.0.0, so it can be closed.
